In this exercise, you will practice basic `dplyr` (the relevant package within `tidyverse`) operations on a dataset of gene expression levels across human tissues (from the GTEx Project). The dataset (`dt`) has one row per gene, with columns:

-   `Name`: the Ensembl gene ID\
-   `Description`: the gene symbol
-   One column per tissue containing the median expression levels (in TPM units)

------------------------------------------------------------------------

### Start by loading the data into R:

Note the arguments, which specify that the first two lines of the file will be skipped, that the file contains a header line that should be used to form column names, and that the columns are tab (`\t`) separated.

```{r}
library(tidyverse)
dt <- read_delim("GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz", skip = 2, col_names = TRUE, delim = "\t")
```

### Q1. How many genes are in the dataset, and how many tissues are represented?

```{r}

# code and answer here

# number of genes
paste(length(unique(dt$Name)))

# number of tissues
length(colnames(dt)) - 2
```

------------------------------------------------------------------------

### Q2. How many genes are expressed above a given threshold in each tissue?

A common threshold is 1 TPM to define whether a gene is "expressed."\
Compute, for each of the three tissues, `Liver`, `Testis`, and `Whole Blood`, how many genes are expressed above this threshold.\
Then identify which tissue has the **most** and which has the **fewest** expressed genes. See this paper for one possible explanation: Xia et al. 2021: [PMC7891839](https://pmc.ncbi.nlm.nih.gov/articles/PMC7891839/).

Do the results change if you set the threshold higher, say at 10 TPM?

```{r}

# Code here

paste('Liver 1tpm: ', length(dt$Name[dt$Liver > 1]))
paste('\nTestis 1tpm: ', length(dt$Name[dt$Testis > 1]))
paste('\nWhole Blood: ', length(dt$Name[dt$`Whole Blood`>1]))


# Interpretation here as a comment
# At 1 tpm,  whole blood < liver < testis

paste('Liver 10tpm: ', length(dt$Name[dt$Liver > 10]))
paste('\nTestis 10tpm: ', length(dt$Name[dt$Testis > 10]))
paste('\nWhole Blood 10tpm: ', length(dt$Name[dt$`Whole Blood`>10]))

# at 10 tpm, whole blood < liver < testis
```

------------------------------------------------------------------------

### Q3. What are the 20 most highly expressed genes in the liver?

What do you notice about many of the genes in this list? Can you propose a hypothesis to explain your observation?

```{r}

# Code here
dt %>%
     arrange(desc(Liver)) %>% select(Description) %>% head(20)

# Interpretation here as a comment
# A lot of these genes are mitochondrial genes. There are a lot of mitochondria in the liver so this makes sense.

```

------------------------------------------------------------------------

### Q4. Compare the mean expression levels of mitochondrial versus nuclear-encoded genes in the Liver.

Genes encoded by the mitochondrial genome have the gene symbol prefix `MT-`. The function `str_starts()` from the `stringr` package in `tidyverse` tests whether a string (or a vector of strings) start with a given pattern, returning TRUE/FALSE (or a boolean vector) as output.

Use `str_starts()` to identify which genes are mitochondrially encoded and store this information in a new Boolean column named `is_mitochondrial`. Next, compute the mean and median expression of mitochondrial genes in the liver, as well as the mean and median expression of nuclear-encoded genes in the liver and compare them. Do mitochondrial genes exhibit higher or lower mean and median expression than nuclear genes? What does the difference between the mean and median tell you about the distribution of gene expression.

```{r}
# Code here
is_mitochondrial <- str_starts(dt$Description, 'MT-')

mean(dt$Liver[is_mitochondrial])
median(dt$Liver[is_mitochondrial])

mean(dt$Liver[!is_mitochondrial])
median(dt$Liver[!is_mitochondrial])
# Interpretation here as a comment
# The mitochondrial genes are more highly expressed in the liver than nuclear genes. The mean of mt gene expression is much higher than the median, which means that about half of mt genes are expressed at near 0 TPM, and the other half are skewed at a very high gene expression.
```

------------------------------------------------------------------------

### Q5. Are liverâ€™s most highly expressed genes also highly expressed in another tissue?

Take the 20 most highly exprssed genes in **Liver** and check how they are expressed in the **Heart - Left Ventricle**. Identify:\
1. one gene that is highly expressed in both tissues\
2. one gene that is highly expressed in Liver but very low in Heart

Look up basic information about these genes on the internet. Do their expression patterns make sense given their known functions?

```{r}

# Code here

liver_genes <- dt %>%
    arrange(desc(Liver)) %>% select(Description) %>% head(20)

dt %>% filter(Description %in% liver_genes$Description) %>% select(c(Description, `Heart - Left Ventricle`))

# Interpretation here as a comment
# ATP6 is highly expressed in both heart and liver, whereas FGB is highly expressed in the liver but not in the heart. FGB encodes fibrinogen, which is a protein used in blood clotting. It is primarily synthesized and secreted in the liver, which is why it would make sense that it is highly expressed in the liver and not in the heart.
```

------------------------------------------------------------------------

### Q6. Comparing gene expression between tissues

Another way to compare tissues is to ask how similar are their expression patterns overall, across all genes?

R has a built-in function called `cor.test()` that measures the correlation between two vectors of numbers. Remember that you can extract the contents of a given column of a tibble as a vector using the syntax `tibble_name$column_name`. If a column name contains spaces, enclose the column name in backticks such as `` tibble_name$`column name` ``.

Use `cor.test()` with the argument `method = "kendall"` to compute: - the Kendall correlation between `Liver` and `Heart - Left Ventricle` - the Spearman correlation between `Liver` and `Brian - Cortex`

Which pair of tissues is more strongly correlated? Does this result make sense biologically?

```{r}
# Liver vs Heart answer
cor.test(dt$Liver, dt$`Heart - Left Ventricle`, method='kendall')
# 0.8053218

# Liver vs Brain answer
cor.test(dt$Liver, dt$`Brain - Cortex`, method='kendall')
# 0.7342544

# interpretation
# The expression of genes in the liver is more closely aligned with those of the heart than in the brain. Given that both the heart and liver are in the cardiovascular system, while the brain is not, I would expect that the heart and liver would have more pathways in common than with the brain.
```

------------------------------------------------------------------------

### Optional: further exploration

What other patterns are present in the data? Use tidyverse functions to explore and summarize the data in different ways and see what you can learn.

```{r}
# answer here
```
